src/iexamine.cpp

```diff
@@ -3142,45 +3139,106 @@ void iexamine::aggie_plant( Character &you, const tripoint_bub_ms &examp )
-    const furn_id &cur_kiln_type = here.furn( examp );
-    furn_id next_kiln_type = furn_str_id::NULL_ID();
-    if( cur_kiln_type == furn_f_kiln_empty ) {
-        next_kiln_type = furn_f_kiln_full;
-    } else if( cur_kiln_type == furn_f_kiln_metal_empty ) {
-        next_kiln_type = furn_f_kiln_metal_full;
-    } else if( cur_kiln_type == furn_f_kiln_portable_empty ) {
-        next_kiln_type = furn_f_kiln_portable_full;
-    } else {
-        debugmsg( "Examined furniture has action kiln_empty, but is of type %s",
-                  cur_kiln_type.id().c_str() );
+    std::string f_name = here.furnname( examp );
+
+    if( !you.has_charges( itype_fire, 1 ) ) {
+        add_msg( _( "This %s is ready to be fired, but you have no fire source." ), f_name );
+        return;
+    } else if( !query_yn( _( "Fire the %s?" ), f_name ) ) {
         return;
     }
 
-    static const std::set<material_id> kilnable{ material_wood, material_bone };
-    bool fuel_present = false;
+    // Determine the fire starter to use.
+    auto is_firestarter = []( const item & it ) {
+        return it.has_flag( flag_FIRESTARTER ) || it.has_flag( flag_FIRE );
+    };
+
+    std::multimap<int, item *> firestarters;
+
+    for( item *it : you.items_with( is_firestarter ) ) {
+        add_firestarter( it, firestarters, you, examp );
+    }
+
+    const bool has_bionic_firestarter = you.has_bionic( bio_lighter ) &&
+                                        you.enough_power_for( bio_lighter );
+
+    bool use_cbm = false;
+
+    // Query to use a CBM over another firestarter only if the firestarters list isn't empty.
+    if( has_bionic_firestarter && !firestarters.empty() ) {
+        use_cbm = query_yn( _( "Use a CBM to start a fire" ) );
+    } else if( has_bionic_firestarter ) {
+        use_cbm = true;
+    }
+
+    // Try to start the smoking rack/kiln.
+    if( !use_cbm ) {
+        for( auto &firestarter : firestarters ) {
+            item *it = firestarter.second;
+            const use_function *usef = it->type->get_use( "firestarter" );
+            const firestarter_actor *actor = dynamic_cast<const firestarter_actor *>( usef->get_actor_ptr() );
+            you.add_msg_if_player( _( "You attempt to start a fire with your %s…" ), it->tname() );
+            const ret_val<void> can_use = actor->can_use( you, *it, examp );
+            if( can_use.success() ) {
+                const int charges = actor->use( &you, *it, examp ).value_or( 0 );
+                you.use_charges( it->typeId(), charges );
+                return;
+            } else {
+                you.add_msg_if_player( m_bad, can_use.str() );
+            }
+        }
+    } else {
+        you.mod_power_level( -bio_lighter->power_activate );
+        you.mod_moves( -to_moves<int>( 1_seconds ) );
+        firestarter_actor::resolve_firestarter_use( &you, &get_map(), examp,
+                st );
+    }
+}
+
+static units::volume total_volume_at_pos( const tripoint_bub_ms &examp )
```

Comment:
i question moving this into a separate function, given that it's only called once. But no strong opinions here, it's probably fine.

Reply:
Used to be called twice, that's why I moved the logic into a function along the way I apparently lost one of the calls...